src/istio/utils/utils.cc

```diff
@@ -0,0 +1,73 @@
+/* Copyright 2018 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/istio/utils/utils.h"
+
+#include <sstream>
+#include <vector>
+
+namespace istio {
+namespace utils {
+
+namespace {
+const std::string kNamespaceKey("ns");
+const char kDelimiter = '/';
+}  // namespace
+
+bool GetSourceNamespace(const std::string& principal,
+                        std::string* source_namespace) {
+  if (source_namespace) {
+    // The namespace is a substring in principal with format:
+    // "<DOMAIN>/ns/<NAMESPACE>/sa/<SERVICE-ACCOUNT>". '/' is not allowed to
+    // appear in actual content except as delimiter between tokens.
+    // The following algorithm extracts the namespace based on the above format
+    // but is a little more flexible. It assumes that the principal begins with
+    // a <DOMAIN> string followed by <key,value> pairs separated by '/'.
+
+    // Skip the first <DOMAIN> string and check remaining <key, value> pairs.
+    size_t begin = principal.find(kDelimiter);
```

Comment:
Does it simpler to find the (first) "/ns/" instead?

Reply:
That would assume the `/ns/` is the first key and would not work for something like this `cluster.local/sa/ns/ns/real_ns`.

Reply:
Not necessarily, you search for the first appearance of `/ns/`, so it still return the position after `sa` in the example above.

Reply:
I'm confused, If I search for the first `/ns/` for the example,
```
            cluster.local/sa/ns/ns/real_ns
first /ns/: ----------------^
ns again    --------------------^
```
then we'll wrongly treat the following `ns` as the namespace which is actually the value of `sa`.

Reply:
Ah I see what you mean, you're only commenting on line 40, I thought you're commenting on the whole algorithm.
But I think it might still have issues, like for this: `cluster.local/ns/abc/sa/user`, I'll have to change the following code to handle this specially.

Reply:
In the function comment, you said the expected format is `"<DOMAIN>/ns/<NAMESPACE>/sa/<SERVICE-ACCOUNT>"`, where DOMAIN cannot have `/`. Do we need to support other format?

Reply:
Yeah, `<DOMAIN>/ns/<NAMESPACE>/sa/<SERVICE-ACCOUNT>` is the current format, I'm trying to make the code more flexible so that it doesn't require the `ns` to be the first one. If you think this is unnecessary, we can simplify the code greatly.